src/ophyd_async/core/_derived_signal.py

```diff
@@ -35,15 +37,27 @@ def __init__(
         self._set_derived = set_derived
         # Check the raw and transform devices match the input arguments of the Transform
         if transform_cls is not Transform:
-            expected = list(transform_cls.model_fields) + [
-                x
-                for x in get_type_hints(transform_cls.raw_to_derived)
-                if x not in ["self", "return"]
-            ]
-            if set(expected) != set(raw_and_transform_devices):
+            # Populate expected parameters and types
+            expected = {
+                **{k: f.annotation for k, f in transform_cls.model_fields.items()},
+                **{
+                    k: v
+                    for k, v in get_type_hints(transform_cls.raw_to_derived).items()
+                    if k not in {"self", "return"}
+                },
+            }
+
+            # Populate received parameters and types
+            # Use Signal datatype, or Locatable datatype, or set type as None
+            received = {
+                k: v.datatype if hasattr(v, "datatype") else get_locatable_type(v)
```

Comment:
Slightly prefer explicit checking on `Signal` type:
```suggestion
                k: v.datatype if isinstance(v, Signal) else get_locatable_type(v)
```

Reply:
Oh yeah, I do prefer this as well! Amended.